src/librustc_codegen_ssa/mir/mod.rs

```diff
@@ -23,10 +23,10 @@ use rustc::mir::traversal;
 use self::operand::{OperandRef, OperandValue};
 
 /// Master context for codegenning from MIR.
-pub struct FunctionCx<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> {
+pub struct FunctionCx<'a, 'b, 'tcx, Bx: BuilderMethods<'a, 'tcx>> {
     instance: Instance<'tcx>,
 
-    mir: &'a mir::Body<'tcx>,
+    mir: &'b mir::Body<'tcx>,
```

Comment:
Making this a distinct lifetime solved a bunch of issues. Otherwise the `Bx` was extending the lifetime on the `&mut` for `BodyCache` and wreaking havoc on my abilty to understand what was going wrong.

Reply:
does every use of `FunctionCx` need a `BodyCache`?

Reply:
Turns out that having a `BodyCache` def of:
```rust
struct BodyCache<'tcx> {
  cache: Cache,
  body: Body<'tcx>,
}
```

Is all we needed to solve this. :smile: